### PR TITLE
Enable the ability to move/resize without clicking

### PR DESCRIPTION
**Pull request by joesteele**
This PR adds the ability to move and resize windows without needing to click first. Pressing the move or resize hotkey combination and then moving your mouse will move or resize the window currenly under the mouse.

Options:
- `enableWithoutClick`: Allows use without clicking (default: `false`)
- `mouseActivationThreshold`: (optional) Threshold of mouse movement before activating when `enableWithoutClick` is `true` (default: `15`)

### CODE REVIEW

File: init.lua [hunk 1]
```diff
@@ -81,14 +81,19 @@ function SkyRocket:new(options)
...
-    dragging = false,
-    dragType = nil,
     moveStartMouseEvent = buttonNameToEventType(options.moveMouseButton or 'left', 'moveMouseButton'),
     moveModifiers = options.moveModifiers or {'cmd', 'shift'},
     windowCanvas = createResizeCanvas(options.opacity or 0.3),
     resizeStartMouseEvent = buttonNameToEventType(options.resizeMouseButton or 'left', 'resizeMouseButton'),
     resizeModifiers = options.resizeModifiers or {'ctrl', 'shift'},
     targetWindow = nil,
+
+    canDrag = false,
```
Comment (joesteele) [inline init.lua:91]: I added this `canDrag` state to represent when dragging has started in mouse-only mode, but the `mouseActivationThreshold` hasn't yet been crossed.

File: init.lua [hunk 2]
```diff
@@ -239,12 +332,6 @@ function SkyRocket:handleClick()
       self.dragging = true
       self.targetWindow = currentWindow
 
-      if isMoving then
-        self.dragType = dragTypes.move
-      else
-        self.dragType = dragTypes.resize
-      end
```
Comment (joesteele) [inline init.lua:246]: I moved the responsibility of setting the `dragType` into `handleFlags`.

File: init.lua [hunk 3]
```diff
@@ -260,4 +347,42 @@ function SkyRocket:handleClick()
...
+        self.targetWindow = currentWindow
+
+        self:resizeCanvasToWindow()
+        self.mouseMoveHandler:start()
+      end
+    else
+      self.canDrag = false
+      if self.enableWithoutClick then
+        self:cancel()
+      end
```
Comment (joesteele) [inline init.lua:381]: This clause would happen when they release the hotkey combo.

**Comment by joesteele** [comment]
Also, @dbalatero, I'm not partial to any of the code I wrote. I'm happy to change anything and of course you're welcome to edit/change things directly as well if you are prefer.

**Comment by Hossman333** [comment]
Nice! I was wondering if this was available. This is a great spoon and adding support for not having to click would be sweet.

**Comment by joesteele** [comment]
It works (I use it every day), but I probably added too much complexity. Every once in awhile the drag/resize state doesn't get unset when releasing the modifier keys. It fixes itself by pressing one of the modifier keys again.

I've been meaning to come back and fix/clean that up and see if I can get it back closer to the original implementation.

Been too busy (or lazy) though since it works well enough for me as-is.

**Comment by Hossman333** [comment]
Awesome. Thanks for the quick reply. Yeah, I'm looking for something like this to replace my better touch tool implementation. To install your branch locally should I just clone your fork and load it up the same way?


```
git clone https://github.com/joesteele/SkyRocket.spoon.git ~/.hammerspoon/Spoons/SkyRocket.spoon
```

Example:
```
local SkyRocket = hs.loadSpoon("SkyRocket")

sky = SkyRocket:new({
  -- Opacity of resize canvas
  opacity = 0.3,

  -- Which modifiers to hold to move a window?
  moveModifiers = {'cmd', 'shift'},

  -- Which mouse button to hold to move a window?
  moveMouseButton = 'left',

  -- Which modifiers to hold to resize a window?
  resizeModifiers = {'ctrl', 'shift'},

  -- Which mouse button to hold to resize a window?
  resizeMouseButton = 'left',
})
```

**Comment by dbalatero** [comment]
@Hossman333 You have to check out their `no-click` branch too; simply cloning the repo will put you on `master` which will not have the changes. You can see this in the PR metadata:

image

**Comment by dbalatero** [comment]
> It works (I use it every day), but I probably added too much complexity. Every once in awhile the drag/resize state doesn't get unset when releasing the modifier keys. It fixes itself by pressing one of the modifier keys again.

Happy for you to have this fork, but I think I'm probably happy with the spoon as it is (for me it's feature complete) and I'm not sure I want to take on reviewing, testing, and maintaining this change. For me, this repo is more or less static and I'm not actively maintaining it (or anything really, I have pretty much no time for GH issue triage and OSS maintenance unfortunately).

**Comment by joesteele** [comment]
Thanks @dbalatero for the awesome spoon and totally get it! I'll close this PR out.

**Pull request closed without merging**